src/rate_limiter.rs

```diff
@@ -316,72 +648,66 @@ mod tests {
+        // A client that arrives late should only be able to put one large message.
+        let late_client = IpAddr::from([255, 255, 255, 255]);
+        assert!(large_message_can_be_added(&mut rate_limiter, &late_client));
+        // And not a byte more.
+        assert!(small_message_cannot_be_added(
+            &mut rate_limiter,
+            &late_client,
+        ));
+
+        // None of the saturated clients should be able to put any more large messages.
+        for client in &clients {
+            assert!(large_message_cannot_be_added(&mut rate_limiter, client));
+        }
+
+        // Now we wait for the remaining part of each saturating client's message to drain.
+        let wait_millis = (num_clients + 2) * MIN_CLIENT_CAPACITY * 100 / RATE as u64;
         FakeClock::advance_time(wait_millis);
-        assert_large_message_can_be_added(&mut rate_limiter, &client1);
-        assert_large_message_cannot_be_added(&mut rate_limiter, &client2);
-
-        // After a further `3 * wait_millis`, the second and third messages should both have drained
-        // allowing both clients to add new large messages.
-        FakeClock::advance_time(3 * wait_millis);
-        assert_large_message_can_be_added(&mut rate_limiter, &client1);
-        assert_large_message_can_be_added(&mut rate_limiter, &client2);
+
+        // Now, the late client should only have had half its message drained but it should
+        // still be able to put another message because of the soft capacity.
+        assert!(large_message_can_be_added(&mut rate_limiter, &late_client));
+
+        // All of the initial clients should be able to put large messages again too.
+        for client in &clients {
+            assert!(large_message_can_be_added(&mut rate_limiter, client));
+        }
     }
 
     /// Checks that many clients can all add messages at the same rate.
     #[test]
     fn many_clients() {
         let mut rate_limiter = RateLimiter::new(false);
-        let num_clients = 100;
+        let num_clients = 50;
         let mut clients_and_counts = (0..num_clients)
             .map(|i| (IpAddr::from([i, i, i, i]), 0))
             .collect::<BTreeMap<_, _>>();
-        let payload = unwrap!(serialisation::serialise(
-            &UserMessage::Request(Request::GetIData {
-                name: rand::random(),
-                msg_id: MessageId::new(),
-            }),
-        ));
-        let hash = sha3_256(&payload);
         let mut rng = SeededRng::thread_rng();
 
         let start = FakeClock::now();
-        for _ in 0..500 {
+        for _ in 0..200 {
             // Each client tries to add a large request and increments its count on success.
             for (client, count) in &mut clients_and_counts {
-                if rate_limiter
-                    .add_message(client, &hash, 1, 0, &payload)
-                    .is_ok()
-                {
+                if add_user_msg_part(&mut rate_limiter, client, &random_payload()).is_ok() {
```

Comment:
I think it's not worth making it exact because we're dealing with an average rate, and a discrete time simulation of a continuous time process. I don't think leaving off the last clock advance is "right" any more than it is "wrong". If there are enough iterations, 1 second shouldn't make a difference... and if we don't want to worry about iterating 10,000 times, I think an off-by-one allowance is OK? Or am I missing something? I did a sketch of the process as I understand it:

https://sketch.io/render/sk-d3dad899d03169b273d562b70b3829e6.jpeg